main.tf

```diff
@@ -171,8 +171,8 @@ resource "azurerm_log_analytics_solution" "main" {
   solution_name         = "ContainerInsights"
   location              = coalesce(var.location, data.azurerm_resource_group.main.location)
   resource_group_name   = var.resource_group_name
-  workspace_resource_id = azurerm_log_analytics_workspace.main[0].id
-  workspace_name        = azurerm_log_analytics_workspace.main[0].name
+  workspace_resource_id = var.log_analytics_workspace == null ? azurerm_log_analytics_workspace.main[0].id : var.log_analytics_workspace.id
```

Comment:
Hi, do we need add this `var.log_analytics_workspace == null` criteria to `azurerm_log_analytics_solution`'s `count` expression, since the log analytics resource is injected by the module caller？

Reply:
`azurerm_log_analytics_solution` is `ContainerInsights` solution connecting newly created AKS and existing (or just created) Workspace. I think that current `count` expression `var.enable_log_analytics_workspace ? 1 : 0` is valid. If caller specifies `enable_log_analytics_workspace = true` then caller either wants to create a new Workspace or use existing. In other words `ContainerInsights` solution is created always when `enable_log_analytics_workspace = true`, but the created solution can be connected to different Workspaces, which is solved later by conditions on `workspace_resource_id` and `workspace_name`.

Reply:
I have a question though, because I checked in `azurerm_log_analytics_solution` docs that:
![image](https://user-images.githubusercontent.com/13100466/177746713-b31ca94e-828f-4204-8add-b4421f3265d4.png)

In my current solution, existing Workspace `var.log_analytics_workspace` must be placed in the same RG as `var.resource_group_name`. Do you think it is valid assumption? It was valid in my use case.

Reply:
> `azurerm_log_analytics_solution` is `ContainerInsights` solution connecting newly created AKS and existing (or just created) Workspace. I think that current `count` expression `var.enable_log_analytics_workspace ? 1 : 0` is valid. If caller specifies `enable_log_analytics_workspace = true` then caller either wants to create a new Workspace or use existing. In other words `ContainerInsights` solution is created always when `enable_log_analytics_workspace = true`, but the created solution can be connected to different Workspaces, which is solved later by conditions on `workspace_resource_id` and `workspace_name`.

My concern is what if the caller has used another module that creates log analytics domain resources? It looks like we need both workspace and solution to make log work, so maybe we should treat them as an inseparable entity?

Reply:
I've verified that the workspace can be put in different rg, but must be in the same rg with solution, so I prefer we add this `var.log_analytics_workspace == null` criteria to solution's `count` expression, leave the choice to the caller.

Reply:
Hi @viters my concern is can our module be composited with other modules easily. When you opened this pr I do agree with you because what if the caller want to use another module to manage log analytics related resources? Your pr makes it possible to avoid a duplicated workspace and leaves the choice to the caller.

So what if the caller has a module that manage all log analytics related resources, even in a standalone repo along with its own state backend? There's no way for the caller to avoid bind the workspace to a solution that he doesn't want.

I think you pr's value is giving callers the right of composition, so can we give the caller more? Thanks!

Reply:
@lonegunmanb 
Oh, I got it. Your idea is to add a way to skip creation of `azurerm_log_analytics_solution` as well! I understand now, I will get to it!

Reply:
@lonegunmanb 
As we discussed, I've added `var.log_analytics_solution_id` and `azurerm_log_analytics_solution` resource will not be created if `var.log_analytics_solution_id == null`.

I've used `string id` instead of `bool` to avoid breaking changes in future - we do not need the solution ID now, but it may be required in future Azure API somewhere, e.g. in `oms_agent` block.